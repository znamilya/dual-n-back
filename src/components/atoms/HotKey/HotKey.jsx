import React, { PropTypes } from 'react';
import bemCl from 'bem-cl';

import './HotKey.styl';


const b = bemCl('hot-key');

class HotKey extends React.Component {

    static propTypes = {
        value: PropTypes.string.isRequired,
        desc: PropTypes.string.isRequired,
        disabled: PropTypes.bool,
        onFire: PropTypes.func.isRequired,
    };

    constructor(props) {
        super(props);

        this.handleDocumentKeyPress = this.handleDocumentKeyPress.bind(this);
    }


    /* ------------------------------------------------------------------------------------------ */
    /* REACT                                                                                      */
    /* ------------------------------------------------------------------------------------------ */
    componentWillReceiveProps(nextProps) {
        if (this.props.disabled && !nextProps.disabled) {
            document.addEventListener('keypress', this.handleDocumentKeyPress);
        }
    }

    componentWillUnmount() {
        document.removeEventListener('keypress', this.handleDocumentKeyPress);
    }


    /* ------------------------------------------------------------------------------------------ */
    /* HANDLERS                                                                                   */
    /* ------------------------------------------------------------------------------------------ */
    handleDocumentKeyPress(e) {
        if (e.key === this.props.value) {
            this.props.onFire();
        }
    }


    /* ------------------------------------------------------------------------------------------ */
    /* RENDER                                                                                     */
    /* ------------------------------------------------------------------------------------------ */
    render() {
        return (
            <div className={b({
                disabled: this.props.disabled,
            })}>
                <div className={b('desc')}>
                    {this.props.desc}
                </div>
                <div className={b('value')}>
                    {this.props.value}
                </div>
            </div>
        );
    }
}


export default HotKey;
