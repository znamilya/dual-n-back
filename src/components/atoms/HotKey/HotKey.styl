.hot-key
    padding: 15px

    text-align: center


    &__value
        font-size: 24px
        text-transform: uppercase
        font-weight: bold
