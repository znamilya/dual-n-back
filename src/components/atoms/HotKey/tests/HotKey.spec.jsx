import React from 'react';

import HotKey from '../HotKey';


describe.only('<HotKey />', () => {
    it('should render itself', () => {
        const wrapper = shallow(
            <HotKey
                value="l"
                desc="Position match"
                onFire={() => {}}
            />
        );

        expect(wrapper).toMatchSnapshot();
    });

    it('should render passed letter', () => {
        const letter = 'a';
        const wrapper = shallow(
            <HotKey
                value="a"
                desc="desc"
                onFire={() => {}}
            />
        )

        expect(wrapper.find('.hot-key__value').text()).toBe(letter);
    });

    it('should disable', () => {
        let wrapper = shallow(
            <HotKey
                value="l"
                desc="Position match"
                disabled
                onFire={() => {}}
            />
        );

        expect(wrapper).toMatchSnapshot();
    });

    describe('onFire', () => {
        const events = {};
        let onFire;

        document.addEventListener = jest.fn((event, cb) => {
            events[event] = cb;
        });

        beforeEach(() => {
            onFire = jest.fn();
        });

        it('should not be called if component disabled', () => {
            const letter = 'l';
            const wrapper = mount(
                <HotKey
                    value={letter}
                    desc="Position match"
                    disabled
                    onFire={onFire}
                />
            );

            events.keypress({ key: letter });

            expect(onFire).not.toHaveBeenCalled();
        });

        it('should not be called when user press different letter', () => {
            const wrapper = mount(
                <HotKey
                    value="l"
                    desc="Position match"
                    onFire={onFire}
                />
            );

            events.keypress({ key: 'k' });

            expect(onFire).not.toHaveBeenCalled();
        });

        it('should be called when user press passed letter', () => {
            const letter = 'l';
            const wrapper = mount(
                <HotKey
                    value={letter}
                    desc="Position match"
                    onFire={onFire}
                />
            );

            events.keypress({ key: letter });

            expect(onFire).toHaveBeenCalled();
        });
    });
});
