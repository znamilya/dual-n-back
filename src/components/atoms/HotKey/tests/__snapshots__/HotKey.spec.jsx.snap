exports[`<HotKey /> should disable 1`] = `
<div
  className="hot-key hot-key_disabled">
  <div
    className="hot-key__desc">
    Position match
  </div>
  <div
    className="hot-key__value">
    l
  </div>
</div>
`;

exports[`<HotKey /> should render itself 1`] = `
<div
  className="hot-key">
  <div
    className="hot-key__desc">
    Position match
  </div>
  <div
    className="hot-key__value">
    l
  </div>
</div>
`;
